exchanges/binance/binance_websocket.go

```diff
@@ -454,66 +455,158 @@ func (b *Binance) UpdateLocalBuffer(wsdp *WebsocketDepthStream) error {
 		return err
 	}
 
-	currencyPair, err := currency.NewPairFromFormattedPairs(wsdp.Pair,
+	currencyPair, err := currency.NewPairFromFormattedPairs(u.Pair,
 		enabledPairs,
 		format)
 	if err != nil {
 		return err
 	}
 
-	currentBook := b.Websocket.Orderbook.GetOrderbook(currencyPair, asset.Spot)
+	return b.CheckAndLoadBook(currencyPair, asset.Spot, u)
+}
+
+// CheckAndLoadBook buffers updates, spawns workers for http book request and
+// lets slip routine, validates buffer contents and returns required work when
+// needed
+func (b *Binance) CheckAndLoadBook(c currency.Pair, ass asset.Item, u *WebsocketDepthStream) error {
+	b.mtx.Lock() // protects fetching book
+	defer b.mtx.Unlock()
+
+	ch, ok := b.pipe[c.String()]
+	if !ok {
+		ch = make(chan *WebsocketDepthStream, 1000)
+		b.pipe[c.String()] = ch
+	}
+	select {
+	case ch <- u:
+	default:
+		return fmt.Errorf("channel blockage %s", c)
+	}
+
+	if b.fetchingbook[c.String()] {
+		return nil
+	}
+
+	return b.BleedPipe(c, ass)
+}
+
+// BleedPipe applies the buffer to the orderbook
+func (b *Binance) BleedPipe(cp currency.Pair, a asset.Item) error {
```

Comment:
I think this would be more accurate to call it something like `applyBufferedUpdated`, but its not very high on the list of things :D